### PR TITLE
Partitioning 2g

### PR DESCRIPTION
**Pull request by drwells**
This is the last partitioning change :tada: 

I want to run some more tests before we merge this, so I am marking it as a work in progress for now.

@cpuelz If you want to give this a spin with the heart model, use this branch and attach the load balancer to the time integrator. Look at how I modified example 4 to see how this is done. You can check the logs to see how it repartitions:

```
workload estimate on processor 0  = 136880 (pid is 19028)
workload estimate on processor 1  = 124576 (pid is 19029)
workload estimate on processor 2  = 126832 (pid is 19030)
workload estimate on processor 3  = 156626 (pid is 19031)
workload estimate on processor 4  = 124834 (pid is 19032)
workload estimate on processor 5  = 143008 (pid is 19033)
workload estimate on processor 6  = 133872 (pid is 19034)
workload estimate on processor 7  = 132770 (pid is 19035)
workload estimate on processor 8  = 128098 (pid is 19036)
workload estimate on processor 9  = 139744 (pid is 19037)
workload estimate on processor 10 = 135072 (pid is 19038)
workload estimate on processor 11 = 128354 (pid is 19039)
workload estimate on processor 12 = 141834 (pid is 19040)
workload estimate on processor 13 = 121366 (pid is 19041)
workload estimate on processor 14 = 129468 (pid is 19042)
workload estimate on processor 15 = 130472 (pid is 19043)
local active DoFs on processor 0  = 177674 (pid is 19028)
local active DoFs on processor 1  = 168746 (pid is 19029)
local active DoFs on processor 2  = 168720 (pid is 19030)
local active DoFs on processor 3  = 200760 (pid is 19031)
local active DoFs on processor 4  = 165180 (pid is 19032)
local active DoFs on processor 5  = 189079 (pid is 19033)
local active DoFs on processor 6  = 178133 (pid is 19034)
local active DoFs on processor 7  = 173842 (pid is 19035)
local active DoFs on processor 8  = 168656 (pid is 19036)
local active DoFs on processor 9  = 185534 (pid is 19037)
local active DoFs on processor 10 = 179810 (pid is 19038)
local active DoFs on processor 11 = 170213 (pid is 19039)
local active DoFs on processor 12 = 189811 (pid is 19040)
local active DoFs on processor 13 = 160696 (pid is 19041)
local active DoFs on processor 14 = 165431 (pid is 19042)
local active DoFs on processor 15 = 139953 (pid is 19043)
```
which is a beautiful partitioning. You will also need to adjust `max_workload_factor` (the value 0.25 seems to work well) so that we can split enough patches.

### CODE REVIEW

**Comment by drwells** [comment]
For comparison: if we comment out the repartitioning and change the estimated work per quadrature point to zero then the workload estimates no longer make sense, but we can see that the parallelization is quite poor by looking at the number of quadrature points per processor:


```
quadrature points on processor 0  = 30048 (pid is 47732)
quadrature points on processor 1  = 94096 (pid is 47733)
quadrature points on processor 2  = 65500 (pid is 47734)
quadrature points on processor 3  = 0 (pid is 47737)
quadrature points on processor 4  = 0 (pid is 47739)
quadrature points on processor 5  = 10835 (pid is 47740)
quadrature points on processor 6  = 114808 (pid is 47741)
quadrature points on processor 7  = 126025 (pid is 47742)
quadrature points on processor 8  = 139401 (pid is 47743)
quadrature points on processor 9  = 137346 (pid is 47744)
quadrature points on processor 10 = 123094 (pid is 47748)
quadrature points on processor 11 = 124629 (pid is 47749)
quadrature points on processor 12 = 48246 (pid is 47750)
quadrature points on processor 13 = 8546 (pid is 47751)
quadrature points on processor 14 = 0 (pid is 47752)
quadrature points on processor 15 = 0 (pid is 47753)
```

i.e., processors 3, 4, 14, and 15 have nothing to do, and most of the work is on processors 7, 8, 9, 10, and 11. We may as well be only running on five processors instead of sixteen.

File: examples/IBFE/explicit/ex4/input2d [hunk 1]
```diff
@@ -4,14 +4,14 @@ RHO = 1.0
...
 // grid spacing parameters
-MAX_LEVELS = 1                                      // maximum number of levels in locally refined grid
-REF_RATIO  = 2                                      // refinement ratio between levels
-N = 64                                              // actual    number of grid cells on coarsest grid level
-NFINEST = (REF_RATIO^(MAX_LEVELS - 1))*N            // effective number of grid cells on finest   grid level
-DX0 = L/N                                           // mesh width on coarsest grid level
-DX  = L/NFINEST                                     // mesh width on finest   grid level
-MFAC = 2.0                                          // ratio of Lagrangian mesh width to Cartesian mesh width
-ELEM_TYPE = "TRI6"                                  // type of element to use for structure discretization
+MAX_LEVELS = 3                                     // maximum number of levels in locally refined grid
```
Comment (boyceg) [inline examples/IBFE/explicit/ex4/input2d]: Let's not change the "default" input files just yet --- instead add in a new file, `input2d.partitioner`?

File: ibtk/src/utilities/HierarchyIntegrator.cpp [hunk 2]
```diff
@@ -558,6 +558,11 @@ void
 HierarchyIntegrator::registerVisItDataWriter(Pointer<VisItDataWriter<NDIM> > visit_writer)
 {
     d_visit_writer = visit_writer;
+    if (d_workload_idx != IBTK::invalid_index)
```
Comment (boyceg) [inline ibtk/src/utilities/HierarchyIntegrator.cpp]: Let's include a run-time option to turn this on/off. Seems fine to me to default to "on".

File: src/IB/IBFEMethod.cpp [hunk 3]
```diff
@@ -1785,7 +1781,19 @@ void IBFEMethod::endDataRedistribution(Pointer<PatchHierarchy<NDIM> > /*hierarch
     // if we are not initialized then there is nothing to do
     if (d_is_initialized)
     {
-        // This is where we will repartition in a future commit
+        // At this point in the code SAMRAI has already redistributed the
```
Comment (boyceg) [inline src/IB/IBFEMethod.cpp]: Should we add a runtime option to turn on/off dynamic repartitioning of libMesh data?

Comment (drwells) [reply src/IB/IBFEMethod.cpp]: Absolutely. I will move the checklist from #396 to here (which included such items).

File: examples/IBFE/explicit/ex4/input2d (hunk 1 above)
Comment (drwells) [reply examples/IBFE/explicit/ex4/input2d]: I only changed this for testing purposes (e.g., printing workloads). I'll get rid of the changes once that is done.

**Comment by drwells** [comment]
Interestingly, adding more patches improves load balancing but wrecks the performance (at least in 2D) of the fluid solver. Lowering `max_workload_factor` to 0.0625 (from 0.25) doubled the amount of time spent in the fluid solver.

**Comment by boyceg** [comment]
I suspect that this will be less of an issue in 3D, but it would be good to have some tests to find out.

**Comment by drwells** [comment]
The remaining items are:
- [ ] Get rid of the changes to the example step.
- [ ] Clean up the logging
- [ ] Make the (graphical) output of workload estimates conditional
- [ ] Fix the issue where logging info doesn't seem to show up

**Comment by drwells** [comment]
Is getting the patch data index out of the variable database actually possible? I am stuck on the current problem:
1. I want to register a variable with the visit data writer in user code by looking it up in the variable database.
2. the visit writer needs the variable name and patch data index.
3. The method that returns the patch data index, `getVariableId`, is protected.

I am going to solve the original problem by adding code to return the workload patch data index from the parent hierarchy integrator, but it seems odd that registering data for output requires access to protected data.

**Comment by boyceg** [comment]
To get the patch data index, you need to have the Variable / VariableContext pair. Your solution sounds like the easiest one to me.

**Comment by drwells** [comment]
I think this is now done, but I have run out of time today to test the new 'when to use the partitioner' logic. I'll verify that everything works at some point over the next few days.

**Comment by drwells** [comment]
I'm now convinced that the database settings are correctly propagated, so this is done.

File: src/IB/IBFEMethod.cpp [hunk 4]
```diff
@@ -1861,6 +1877,7 @@ IBFEMethod::putToDatabase(Pointer<Database> db)
     db->putBool("d_split_tangential_force", d_split_tangential_force);
     db->putBool("d_use_jump_conditions", d_use_jump_conditions);
     db->putBool("d_use_consistent_mass_matrix", d_use_consistent_mass_matrix);
+    db->putString("d_libmesh_partitioner_type", enum_to_string<LibmeshPartitionerType>(d_libmesh_partitioner_type));
```
Comment (drwells) [inline src/IB/IBFEMethod.cpp]: I forgot about force pushing the change that added this. This is the only change I made since Friday.

**Comment by boyceg** [comment]
Shall we merge?

**Comment by drwells** [comment]
I'm doing some last-minute checks. I'll merge when they are done.

**Comment by drwells** [comment]
This changes some results at the level of roundoff, but I think this is within the bounds of the linear solver tolerance. Lets merge!

**Comment by boyceg** [comment]
What causes the changes? There should be no effect unless the load balance is explicitly enabled?

**Comment by drwells** [comment]
I checked with the load balancer explicitly enabled vs. not enabled (i.e., master).

**Pull request merged**